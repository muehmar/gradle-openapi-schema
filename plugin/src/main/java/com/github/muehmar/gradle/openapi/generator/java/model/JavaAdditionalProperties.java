package com.github.muehmar.gradle.openapi.generator.java.model;

import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaAnyType.javaAnyType;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumContentBuilder;
import com.github.muehmar.gradle.openapi.generator.java.generator.enumpojo.EnumGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaEnumType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.AdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
public class JavaAdditionalProperties {
  private static final JavaName MAP_PROPERTY_NAME = JavaName.fromString("additionalProperties");
  boolean allowed;
  JavaType type;

  public static JavaAdditionalProperties wrap(
      AdditionalProperties additionalProperties, TypeMappings typeMappings) {
    return new JavaAdditionalProperties(
        additionalProperties.isAllowed(),
        JavaType.wrap(additionalProperties.getType(), typeMappings));
  }

  public static JavaAdditionalProperties anyTypeAllowed() {
    return new JavaAdditionalProperties(true, javaAnyType());
  }

  public static JavaAdditionalProperties allowedFor(JavaType type) {
    return new JavaAdditionalProperties(true, type);
  }

  public static JavaAdditionalProperties notAllowed() {
    return new JavaAdditionalProperties(false, javaAnyType());
  }

  public static JavaName additionalPropertiesName() {
    return MAP_PROPERTY_NAME;
  }

  public boolean isNotAllowed() {
    return not(allowed);
  }

  public boolean isValueAnyType() {
    return type.equals(javaAnyType());
  }

  public boolean isNotValueAnyType() {
    return not(type.equals(javaAnyType()));
  }

  public TechnicalPojoMember asTechnicalPojoMember() {
    return TechnicalPojoMember.additionalProperties(type);
  }

  public Optional<EnumGenerator.EnumContent> asEnumContent() {
    final Function<JavaEnumType, Optional<EnumGenerator.EnumContent>> createEnumContent =
        enumType ->
            Optional.of(
                EnumContentBuilder.create()
                    .className(JavaName.fromName(type.getQualifiedClassName().getClassName()))
                    .description("Additional property enum")
                    .members(enumType.getMembers())
                    .build());
    return type.fold(
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        createEnumContent,
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty(),
        ignore -> Optional.empty());
  }
}
