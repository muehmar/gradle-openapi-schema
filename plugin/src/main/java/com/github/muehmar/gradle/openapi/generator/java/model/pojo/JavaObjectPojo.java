package com.github.muehmar.gradle.openapi.generator.java.model.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberContentBuilder.fullMemberContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherContentBuilder.fullWitherContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ConstructorContentBuilder.fullConstructorContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsContentBuilder.fullEqualsContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeContentBuilder.fullHashCodeContentBuilder;
import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringContentBuilder.fullToStringContentBuilder;
import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.NonEmptyList;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.exception.OpenApiGeneratorException;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.WitherGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.EqualsGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.PojoConstructorGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.ToStringGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.PojoType;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAllOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaAnyOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.composition.JavaOneOfComposition;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaPojoName;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.AnyOfContainer;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.auxiliaryy.OneOfContainer;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.name.SchemaName;
import com.github.muehmar.gradle.openapi.generator.model.pojo.ObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import io.github.muehmar.codegenerator.java.JavaModifier;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@PojoBuilder(packagePrivateBuilder = true)
@EqualsAndHashCode
@ToString
public class JavaObjectPojo implements JavaPojo {
  private final JavaPojoName name;
  private final SchemaName schemaName;
  private final String description;
  private final PList<JavaPojoMember> members;
  private final Optional<JavaAllOfComposition> allOfComposition;
  private final Optional<JavaOneOfComposition> oneOfComposition;
  private final Optional<JavaAnyOfComposition> anyOfComposition;
  private final PojoType type;
  private final PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties;
  private final JavaAdditionalProperties additionalProperties;
  private final Constraints constraints;

  JavaObjectPojo(
      JavaPojoName name,
      SchemaName schemaName,
      String description,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition,
      PojoType type,
      PList<JavaRequiredAdditionalProperty> requiredAdditionalProperties,
      JavaAdditionalProperties additionalProperties,
      Constraints constraints) {
    this.name = name;
    this.schemaName = schemaName;
    this.description = Optional.ofNullable(description).orElse("");
    this.members = members;
    this.allOfComposition = allOfComposition;
    this.oneOfComposition = oneOfComposition;
    this.anyOfComposition = anyOfComposition;
    this.type = type;
    this.requiredAdditionalProperties = requiredAdditionalProperties;
    this.additionalProperties = additionalProperties;
    this.constraints = constraints;
    assertPropertiesHaveNotSameNameAndDifferentAttributes(
        name, members, allOfComposition, oneOfComposition, anyOfComposition);
  }

  private static void assertPropertiesHaveNotSameNameAndDifferentAttributes(
      JavaPojoName name,
      PList<JavaPojoMember> members,
      Optional<JavaAllOfComposition> allOfComposition,
      Optional<JavaOneOfComposition> oneOfComposition,
      Optional<JavaAnyOfComposition> anyOfComposition) {
    final PList<JavaPojoMember> allMembers =
        members
            .concat(allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty))
            .concat(oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty))
            .concat(anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty));

    final PList<JavaPojoMember> invalidMembers =
        allMembers
            .filter(
                member1 ->
                    allMembers.exists(member2 -> sameNameButDifferentAttributes(member1, member2)))
            .duplicates(Comparator.comparing(m -> m.getName().asString()));

    if (invalidMembers.nonEmpty()) {
      throw new OpenApiGeneratorException(
          String.format(
              "Cannot create DTO %s: Two or more properties (defined either in properties or in one of the schemas used "
                  + "for a composition (allOf, oneOf or anyOf) have the same name but different attributes (e.g. type, "
                  + "constraints)! Invalid properties: [%s].",
              name, invalidMembers.map(JavaPojoMember::getName).mkString(", ")));
    }
  }

  private static boolean sameNameButDifferentAttributes(JavaPojoMember m1, JavaPojoMember m2) {
    return m1.getName().equals(m2.getName())
        && not(m1.asObjectMember().equals(m2.asObjectMember()));
  }

  public static JavaPojoWrapResult wrap(ObjectPojo objectPojo, TypeMappings typeMappings) {
    if (objectPojo.containsNoneDefaultPropertyScope()) {
      return JavaPojoWrapResultBuilder.create()
          .defaultPojo(createForType(objectPojo, typeMappings, PojoType.DEFAULT))
          .andAllOptionals()
          .requestPojo(createForType(objectPojo, typeMappings, PojoType.REQUEST))
          .responsePojo(createForType(objectPojo, typeMappings, PojoType.RESPONSE))
          .build();
    } else {
      return JavaPojoWrapResult.ofDefaultPojo(
          createForType(objectPojo, typeMappings, PojoType.DEFAULT));
    }
  }

  private static JavaObjectPojo createForType(
      ObjectPojo objectPojo, TypeMappings typeMappings, PojoType type) {
    final PList<JavaPojoMember> members =
        objectPojo
            .getMembers()
            .filter(member -> type.includesPropertyScope(member.getPropertyScope()))
            .map(member -> JavaPojoMember.wrap(member, typeMappings));
    final JavaAdditionalProperties javaAdditionalProperties =
        JavaAdditionalProperties.wrap(objectPojo.getAdditionalProperties(), typeMappings);
    return new JavaObjectPojo(
        JavaPojoName.fromPojoName(type.mapName(objectPojo.getName().getPojoName())),
        objectPojo.getName().getSchemaName(),
        objectPojo.getDescription(),
        members,
        objectPojo
            .getAllOfComposition()
            .map(comp -> JavaAllOfComposition.wrap(comp, type, typeMappings)),
        objectPojo
            .getOneOfComposition()
            .map(
                comp ->
                    JavaOneOfComposition.wrap(
                        comp, objectPojo.getDiscriminator(), type, typeMappings)),
        objectPojo
            .getAnyOfComposition()
            .map(comp -> JavaAnyOfComposition.wrap(comp, type, typeMappings)),
        type,
        objectPojo
            .getRequiredAdditionalProperties()
            .map(
                propName ->
                    JavaRequiredAdditionalProperty.fromNameAndType(
                        propName, javaAdditionalProperties.getType())),
        javaAdditionalProperties,
        objectPojo.getConstraints());
  }

  @Override
  public JavaName getSchemaName() {
    return JavaName.fromName(schemaName.asName());
  }

  @Override
  public JavaName getClassName() {
    return name.asJavaName();
  }

  public JavaPojoName getJavaPojoName() {
    return name;
  }

  public JavaName prefixedClassNameForMethod(String prefix) {
    return name.asJavaName().prefixedMethodName(prefix);
  }

  @Override
  public String getDescription() {
    return description;
  }

  @Override
  public PojoType getType() {
    return type;
  }

  public boolean isSimpleMapPojo() {
    return members.isEmpty()
        && not(allOfComposition.isPresent())
        && not(oneOfComposition.isPresent())
        && not(anyOfComposition.isPresent());
  }

  public boolean hasRequiredMembers() {
    return members.exists(JavaPojoMember::isRequired);
  }

  public boolean hasNotRequiredMembers() {
    return not(hasRequiredMembers());
  }

  public PList<JavaPojoMember> getMembers() {
    return members;
  }

  /** This includes also any possible required additional property. */
  public int getRequiredMemberCount() {
    return members.filter(JavaPojoMember::isRequired).size() + requiredAdditionalProperties.size();
  }

  public PList<JavaPojoMember> getAllMembersForComposition() {
    return getMembers()
        .map(member -> member.asInnerEnumOf(getClassName()))
        .concat(getComposedMembers());
  }

  public PList<JavaPojoMember> getComposedMembers() {
    final PList<JavaPojoMember> allOfMembers =
        allOfComposition.map(JavaAllOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> oneOfMembers =
        oneOfComposition.map(JavaOneOfComposition::getMembers).orElseGet(PList::empty);
    final PList<JavaPojoMember> anyOfMembers =
        anyOfComposition.map(JavaAnyOfComposition::getMembers).orElseGet(PList::empty);
    return allOfMembers.concat(oneOfMembers).concat(anyOfMembers).distinct(Function.identity());
  }

  public PList<JavaPojoMember> getAllMembers() {
    return members.concat(getComposedMembers()).distinct(Function.identity());
  }

  public PList<TechnicalPojoMember> getTechnicalMembers() {
    return getAllMembers()
        .flatMap(JavaPojoMember::getTechnicalMembers)
        .add(additionalProperties.asTechnicalPojoMember());
  }

  public Optional<JavaAllOfComposition> getAllOfComposition() {
    return allOfComposition;
  }

  public Optional<JavaOneOfComposition> getOneOfComposition() {
    return oneOfComposition;
  }

  public PList<JavaObjectPojo> getOneOfPojos() {
    return oneOfComposition
        .map(JavaOneOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasOneOfComposition() {
    return oneOfComposition.isPresent();
  }

  public Optional<JavaAnyOfComposition> getAnyOfComposition() {
    return anyOfComposition;
  }

  public PList<JavaObjectPojo> getAnyOfPojos() {
    return anyOfComposition
        .map(JavaAnyOfComposition::getPojos)
        .map(NonEmptyList::toPList)
        .orElseGet(PList::empty);
  }

  public boolean hasAnyOfComposition() {
    return anyOfComposition.isPresent();
  }

  public PList<JavaRequiredAdditionalProperty> getRequiredAdditionalProperties() {
    return requiredAdditionalProperties;
  }

  public JavaAdditionalProperties getAdditionalProperties() {
    return additionalProperties;
  }

  public Constraints getConstraints() {
    return constraints;
  }

  public MemberGenerator.MemberContent getMemberContent() {
    return fullMemberContentBuilder()
        .isArrayPojo(false)
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .additionalProperties(additionalProperties)
        .build();
  }

  public HashCodeGenerator.HashCodeContent getHashCodeContent() {
    return fullHashCodeContentBuilder().technicalPojoMembers(getTechnicalMembers()).build();
  }

  public EqualsGenerator.EqualsContent getEqualsContent() {
    return fullEqualsContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public ToStringGenerator.ToStringContent getToStringContent() {
    return fullToStringContentBuilder()
        .className(getClassName())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public PojoConstructorGenerator.ConstructorContent getConstructorContent() {
    return fullConstructorContentBuilder()
        .isArray(false)
        .className(getClassName())
        .members(getAllMembers().flatMap(JavaPojoMember::getTechnicalMembers))
        .modifier(Optional.of(JavaModifier.PUBLIC).filter(ignore -> not(isSimpleMapPojo())))
        .additionalProperties(additionalProperties)
        .build();
  }

  public WitherGenerator.WitherContent getWitherContent() {
    return fullWitherContentBuilder()
        .className(getClassName())
        .membersForWithers(getMembers())
        .technicalPojoMembers(getTechnicalMembers())
        .build();
  }

  public Optional<OneOfContainer> getOneOfContainer() {
    return oneOfComposition.map(composition -> new OneOfContainer(name, composition));
  }

  public Optional<AnyOfContainer> getAnyOfContainer() {
    return anyOfComposition.map(composition -> new AnyOfContainer(name, composition));
  }

  @Override
  public <T> T fold(
      Function<JavaArrayPojo, T> onArrayPojo,
      Function<JavaEnumPojo, T> onEnumPojo,
      Function<JavaObjectPojo, T> onObjectPojo) {
    return onObjectPojo.apply(this);
  }
}
