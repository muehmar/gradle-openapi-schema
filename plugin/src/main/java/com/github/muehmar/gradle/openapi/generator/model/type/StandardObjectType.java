package com.github.muehmar.gradle.openapi.generator.model.type;

import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.Type;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.name.PojoName;
import com.github.muehmar.gradle.openapi.generator.settings.PojoNameMapping;
import java.util.Optional;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

/** Could either represent an array type or an object type with properties. */
@EqualsAndHashCode
@ToString
public class StandardObjectType implements ObjectType {
  private final PojoName name;
  private final Nullability nullability;

  private StandardObjectType(PojoName name, Nullability nullability) {
    this.name = name;
    this.nullability = nullability;
  }

  public static StandardObjectType ofName(PojoName name) {
    return new StandardObjectType(name, Nullability.NOT_NULLABLE);
  }

  public PojoName getName() {
    return name;
  }

  @Override
  public StandardObjectType applyMapping(PojoNameMapping pojoNameMapping) {
    return new StandardObjectType(pojoNameMapping.map(name), nullability);
  }

  @Override
  public Type makeNullable() {
    return withNullability(Nullability.NULLABLE);
  }

  @Override
  public StandardObjectType withNullability(Nullability nullability) {
    return new StandardObjectType(name, nullability);
  }

  @Override
  public Optional<EnumObjectType> asEnumObjectType() {
    return Optional.empty();
  }

  @Override
  public Constraints getConstraints() {
    return Constraints.empty();
  }

  @Override
  public Nullability getNullability() {
    return nullability;
  }

  @Override
  public <T> T fold(
      Function<NumericType, T> onNumericType,
      Function<IntegerType, T> onIntegerType,
      Function<StringType, T> onStringType,
      Function<ArrayType, T> onArrayType,
      Function<BooleanType, T> onBooleanType,
      Function<ObjectType, T> onObjectType,
      Function<EnumType, T> onEnumType,
      Function<MapType, T> onMapType,
      Function<AnyType, T> onAnyType,
      Function<MultiType, T> onMultiType) {
    return onObjectType.apply(this);
  }
}
