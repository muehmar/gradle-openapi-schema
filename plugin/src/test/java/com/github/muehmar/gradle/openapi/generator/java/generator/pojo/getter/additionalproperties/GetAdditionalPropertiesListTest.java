package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.additionalproperties;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.additionalproperties.GetAdditionalPropertiesList.getAdditionalPropertiesListGenerator;
import static com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes.stringListType;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NOT_NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.model.name.PojoNames.pojoName;
import static com.github.muehmar.gradle.openapi.generator.settings.ClassTypeMappings.STRING_MAPPING_WITH_CONVERSION;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.assertEquals;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaObjectType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaStringType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaTypes;
import com.github.muehmar.gradle.openapi.generator.model.type.StandardObjectType;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class GetAdditionalPropertiesListTest {
  private Expect expect;

  @Test
  void generate_when_additionalPropertiesNotAllowed_then_noOutput() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties = JavaAdditionalProperties.notAllowed();
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    assertEquals("", writer.asString());
  }

  @Test
  @SnapshotName("notNullableAnyType")
  void generate_when_notNullableAnyType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(JavaTypes.anyType().withNullability(NOT_NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("nullableAnyType")
  void generate_when_nullableAnyType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(JavaTypes.anyType().withNullability(NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("notNullableObjectType")
  void generate_when_notNullableObjectType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(
            JavaObjectType.wrap(
                    StandardObjectType.ofName(pojoName("Object", "Dto")), TypeMappings.empty())
                .withNullability(NOT_NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("nullableObjectType")
  void generate_when_nullableObjectType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(
            JavaObjectType.wrap(
                    StandardObjectType.ofName(pojoName("Object", "Dto")), TypeMappings.empty())
                .withNullability(NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("notNullableListType")
  void generate_when_notNullableListType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(stringListType().withNullability(NOT_NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("nullableListType")
  void generate_when_nullableListType_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(stringListType().withNullability(NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("notNullableStringTypeWithMapping")
  void generate_when_notNullableStringTypeWithMapping_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final TypeMappings stringTypeMapping =
        TypeMappings.ofClassTypeMappings(STRING_MAPPING_WITH_CONVERSION);
    final JavaStringType stringType = JavaStringType.wrap(StringType.noFormat(), stringTypeMapping);

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(stringType.withNullability(NOT_NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("nullableStringTypeWithMapping")
  void generate_when_nullableStringTypeWithMapping_then_matchSnapshot() {
    final Generator<JavaObjectPojo, PojoSettings> generator =
        getAdditionalPropertiesListGenerator();

    final TypeMappings stringTypeMapping =
        TypeMappings.ofClassTypeMappings(STRING_MAPPING_WITH_CONVERSION);
    final JavaStringType stringType = JavaStringType.wrap(StringType.noFormat(), stringTypeMapping);

    final JavaAdditionalProperties additionalProperties =
        JavaAdditionalProperties.allowedFor(stringType.withNullability(NULLABLE));
    final JavaObjectPojo pojo =
        JavaPojos.objectPojo().withAdditionalProperties(additionalProperties);

    final Writer writer = generator.generate(pojo, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
