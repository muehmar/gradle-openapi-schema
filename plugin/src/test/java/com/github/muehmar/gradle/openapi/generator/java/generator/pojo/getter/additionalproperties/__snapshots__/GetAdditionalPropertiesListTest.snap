notNullableAnyType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.AdditionalProperty
java.util.List
java.util.stream.Collectors

@JsonIgnore
public List<AdditionalProperty<Object>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .map(entry -> new AdditionalProperty<>(entry.getKey(), entry.getValue()))
      .collect(Collectors.toList());
}
]


notNullableListType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.AdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<AdditionalProperty<List<String>>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .map(val -> new AdditionalProperty<>(entry.getKey(), val))
                  .map(Stream::of)
                  .orElseGet(Stream::empty))
      .collect(Collectors.toList());
}
]


notNullableObjectType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.AdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<AdditionalProperty<ObjectDto>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .map(val -> new AdditionalProperty<>(entry.getKey(), val))
                  .map(Stream::of)
                  .orElseGet(Stream::empty))
      .collect(Collectors.toList());
}
]


notNullableStringTypeWithMapping=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.AdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<AdditionalProperty<CustomString>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .map(val -> CustomString.fromString(val))
                  .map(val -> new AdditionalProperty<>(entry.getKey(), val))
                  .map(Stream::of)
                  .orElseGet(Stream::empty))
      .collect(Collectors.toList());
}
]


nullableAnyType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.NullableAdditionalProperty
java.util.List
java.util.stream.Collectors

@JsonIgnore
public List<NullableAdditionalProperty<Object>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .map(entry -> NullableAdditionalProperty.ofNullable(entry.getKey(), entry.getValue()))
      .collect(Collectors.toList());
}
]


nullableListType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.NullableAdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<NullableAdditionalProperty<List<String>>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .onValue(val -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), val)))
                  .onNull(() -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), null)))
                  .onAbsent(Stream::empty))
      .collect(Collectors.toList());
}
]


nullableObjectType=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.NullableAdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<NullableAdditionalProperty<ObjectDto>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .onValue(val -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), val)))
                  .onNull(() -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), null)))
                  .onAbsent(Stream::empty))
      .collect(Collectors.toList());
}
]


nullableStringTypeWithMapping=[
com.fasterxml.jackson.annotation.JsonIgnore
com.github.muehmar.openapi.util.NullableAdditionalProperty
java.util.List
java.util.stream.Collectors
java.util.stream.Stream

@JsonIgnore
public List<NullableAdditionalProperty<CustomString>> getAdditionalProperties() {
  return additionalProperties.entrySet().stream()
      .flatMap(
          entry -> 
              castAdditionalProperty(entry.getValue())
                  .map(val -> CustomString.fromString(val))
                  .onValue(val -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), val)))
                  .onNull(() -> Stream.of(NullableAdditionalProperty.ofNullable(entry.getKey(), null)))
                  .onAbsent(Stream::empty))
      .collect(Collectors.toList());
}
]